### PR TITLE
Explain origin of implicit `Sized` obligations and provide suggestions when possible

### PR DESCRIPTION
**Pull request by estebank**
Built on top of #85799. Only last two commits are relevant.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/assoc.rs [hunk 1]
```diff
@@ -58,9 +58,44 @@ impl AssocItem {
...
+                            kind: hir::ImplItemKind::TyAlias(ty),
+                            ..
+                        }),
+                    ) => {
+                        format!(" = {:?}", ty)
+                    }
+                    _ => String::new(),
+                };
+                format!("type {}{};", self.ident, default)
+            }
```
Comment (estebank) [inline compiler/rustc_middle/src/ty/assoc.rs:78]: This doesn't have a test. It wasn't meant to be in this commit 🤦

**Comment by estebank** [comment]
@bors try @rust-timer queue

This should likely not be merged as is: the approach of signaling on the predicate isn't great.

**Comment by tlyu** [comment]
I'm sorry if this is completely off base; I am not completely certain about my understanding of AST lowering and HIR.

> ```
> thread 'rustc' panicked at 'assertion failed: map[k].is_none()', compiler/rustc_middle/src/hir/map/collector.rs:50:5
> ```

To me, this looked like a HIR invariant being violated. I think the `?Sized` bound is being duplicated during AST lowering without getting fresh `HirId` values. Unfortunately, fixing that is somewhat complicated because it isn't a leaf node that's being lowered, so I think we would have to recursively get fresh `HirId` values?

Also, I think the existing code that moves the `?Sized` bound from the `where` clause to the generic parameters is already fairly complicated and fragile. Adding complexity on top of that seems risky to me. I think it would be better to rethink how the trait bound predicates are being collected during type checking, but that's admittedly also difficult.

On a hunch, I dumped the `-Zunpretty=hir-tree` from a simple test file:
```
struct A<T>(&T) where T: ?Sized;
```
and got something including this (note the repeated `HirId` values (`local_id` = 7 or 8) in the `Path` and `PathSegment` inside the `PolyTraitRef` values; also sorry I don't know how to highlight lines within code blocks):
```
Generics {
    params: [
        GenericParam {
```

<details>

```
            hir_id: HirId {
                owner: DefId(0:3 ~ lib[317d]::A),
                local_id: 9,
            },
            name: Plain(
                T#0,
            ),
```
</details>

```
            bounds: [
                Trait(
                    PolyTraitRef {
                        bound_generic_params: [],
                        trait_ref: TraitRef {
                            path: Path {
```

<details>

```
                                span: lib.rs:1:27: 1:32 (#0),
                                res: Def(
                                    Trait,
                                    DefId(2:2854 ~ core[1d1a]::marker::Sized),
                                ),
```
</details>

```
                                segments: [
                                    PathSegment {
                                        ident: Sized#0,
                                        hir_id: Some(
                                            HirId {
                                                owner: DefId(0:3 ~ lib[317d]::A),
                                                local_id: 7,
                                            },
                                        ),
```

<details>

```
                                        res: Some(
                                            Def(
                                                Trait,
                                                DefId(2:2854 ~ core[1d1a]::marker::Sized),
                                            ),
                                        ),
                                        args: None,
                                        infer_args: false,
```
</details>

```
                                    },
                                ],
                            },
                            hir_ref_id: HirId {
                                owner: DefId(0:3 ~ lib[317d]::A),
                                local_id: 8,
                            },
                        },
                        span: lib.rs:1:26: 1:32 (#0),
                    },
                    Maybe,
                ),
            ],
```

<details>

```
            span: lib.rs:1:10: 1:11 (#0),
            pure_wrt_drop: false,
            kind: Type {
                default: None,
                synthetic: None,
            },
```
</details>

```
        },
    ],
    where_clause: WhereClause {
        predicates: [
            BoundPredicate(
                WhereBoundPredicate {
```

<details>

```
                    span: lib.rs:1:23: 1:32 (#0),
                    bound_generic_params: [],
                    bounded_ty: Ty {
                        hir_id: HirId {
                            owner: DefId(0:3 ~ lib[317d]::A),
                            local_id: 10,
                        },
                        kind: Path(
                            Resolved(
                                None,
                                Path {
                                    span: lib.rs:1:23: 1:24 (#0),
                                    res: Def(
                                        TyParam,
                                        DefId(0:5 ~ lib[317d]::A::T),
                                    ),
                                    segments: [
                                        PathSegment {
                                            ident: T#0,
                                            hir_id: Some(
                                                HirId {
                                                    owner: DefId(0:3 ~ lib[317d]::A),
                                                    local_id: 11,
                                                },
                                            ),
                                            res: Some(
                                                Def(
                                                    TyParam,
                                                    DefId(0:5 ~ lib[317d]::A::T),
                                                ),
                                            ),
                                            args: None,
                                            infer_args: false,
                                        },
                                    ],
                                },
                            ),
                        ),
                        span: lib.rs:1:23: 1:24 (#0),
                    },
```
</details>

```
                    bounds: [
                        Trait(
                            PolyTraitRef {
                                bound_generic_params: [],
                                trait_ref: TraitRef {
                                    path: Path {
```

<details>

```
                                        span: lib.rs:1:27: 1:32 (#0),
                                        res: Def(
                                            Trait,
                                            DefId(2:2854 ~ core[1d1a]::marker::Sized),
                                        ),
```
</details>

```
                                        segments: [
                                            PathSegment {
                                                ident: Sized#0,
                                                hir_id: Some(
                                                    HirId {
                                                        owner: DefId(0:3 ~ lib[317d]::A),
                                                        local_id: 7,
                                                    },
                                                ),
```

<details>

```
                                                res: Some(
                                                    Def(
                                                        Trait,
                                                        DefId(2:2854 ~ core[1d1a]::marker::Sized),
                                                    ),
                                                ),
                                                args: None,
                                                infer_args: false,
```
</details>

```
                                            },
                                        ],
                                    },
                                    hir_ref_id: HirId {
                                        owner: DefId(0:3 ~ lib[317d]::A),
                                        local_id: 8,
                                    },
                                },
                                span: lib.rs:1:26: 1:32 (#0),
                            },
                            Maybe,
                        ),
                    ],
```

<details>

```
                },
            ),
        ],
        span: lib.rs:1:17: 1:32 (#0),
    },
```
</details>

**Comment by tlyu** [comment]
Also, I confirmed a build failure for the stage1 `library/std` with
> f33886720db Suggest removal of `?Sized` bound when appropriate

and success with
> 6fdf73ff695 Fix rebase

Same assertion failure:
```
thread 'rustc' panicked at 'assertion failed: map[k].is_none()', compiler/rustc_middle/src/hir/map/collector.rs:50:5
```
during
```
rustc --crate-name core --edition=2018 library/core/src/lib.rs …
```

and it looks like f33886720db did change some relevant AST lowering code.

**Comment by estebank** [comment]
Closing as https://github.com/rust-lang/rust/pull/85799 has all the commits here, I'll try to land everything together, and split it if asked by reviewer.

**Pull request closed without merging**